### PR TITLE
[heft-typescript@next] Add "useTranspilerWorker" option

### PR DESCRIPTION
**Pull request by dmichon-msft**
## Summary
Adds a new option in `typescript.json`, `useTranspilerWorker`, that if `true`, and the tsconfig has `isolatedModules: true`, will split transpilation into a worker thread and run it in parallel with the type checker.

## Details
If opted in and the tsconfig is compatible, will start a worker thread and hand it the list of changed source files and the emit options. Multi-emit then occurs only in the worker, since declaration files are only emitted for the primary module kind.

## How it was tested
Local runs on projects in various configurations. Still needs more testing against large projects with `isolatedModules: true` to determine if the performance is significantly improved.
Verified in both watch and non-watch mode.
Verified using project references.

## Impacted documentation
Documentation for `typescript.json`.

### CODE REVIEW

File: heft-plugins/heft-typescript-plugin/src/TranspilerWorker.ts [hunk 1]
```diff
@@ -0,0 +1,83 @@
+import { parentPort, workerData } from 'node:worker_threads';
```
Comment (iclanton) [inline heft-plugins/heft-typescript-plugin/src/TranspilerWorker.ts:1]:
```suggestion
// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import { parentPort, workerData } from 'node:worker_threads';
```

File: heft-plugins/heft-typescript-plugin/src/TranspilerWorker.ts [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+
+import * as TTypescript from 'typescript';
```
Comment (iclanton) [inline heft-plugins/heft-typescript-plugin/src/TranspilerWorker.ts:3]:
```suggestion
import type * as TTypeScript from 'typescript';
```

File: heft-plugins/heft-typescript-plugin/src/TranspilerWorker.ts [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+  ITranspilationRequestMessage,
+  ITranspilationResponseMessage,
+  ITypescriptWorkerData
+} from './types';
+import type { ExtendedTypeScript } from './internalTypings/TypeScriptInternals';
+import { configureProgramForMultiEmit } from './configureProgramForMultiEmit';
+
+const typedWorkerData: ITypescriptWorkerData = workerData;
+
+const ts: ExtendedTypeScript = require(typedWorkerData.typeScriptToolPath);
```
Comment (iclanton) [inline heft-plugins/heft-typescript-plugin/src/TranspilerWorker.ts:14]: Is this how we're grabbing TypeScript elsewhere?

File: heft-plugins/heft-typescript-plugin/src/TranspilerWorker.ts [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+
+function handleMessage(message: ITranspilationRequestMessage | false): void {
+  if (!message) {
+    process.exit(0);
```
Comment (iclanton) [inline heft-plugins/heft-typescript-plugin/src/TranspilerWorker.ts:18]: `process.exit` just terminates the thread? Interesting...

File: heft-plugins/heft-typescript-plugin/src/configureProgramForMultiEmit.ts [hunk 5]
```diff
@@ -0,0 +1,181 @@
+import type * as TTypescript from 'typescript';
```
Comment (iclanton) [inline heft-plugins/heft-typescript-plugin/src/configureProgramForMultiEmit.ts:1]:
```suggestion
// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import type * as TTypeScript from 'typescript';
```

File: heft-plugins/heft-typescript-plugin/src/types.ts [hunk 6]
```diff
@@ -0,0 +1,50 @@
+import type * as TTypescript from 'typescript';
```
Comment (iclanton) [inline heft-plugins/heft-typescript-plugin/src/types.ts:1]:
```suggestion
// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import type * as TTypeScript from 'typescript';
```

File: heft-plugins/heft-typescript-plugin/src/TypeScriptBuilder.ts [hunk 7]
```diff
@@ -4,40 +4,29 @@
 import * as crypto from 'crypto';
 import * as path from 'path';
 import * as semver from 'semver';
-import type * as TTypescript from 'typescript';
+import * as TTypescript from 'typescript';
```
Comment (iclanton) [inline heft-plugins/heft-typescript-plugin/src/TypeScriptBuilder.ts:7]:
```suggestion
import type * as TTypeScript from 'typescript';
```

File: heft-plugins/heft-typescript-plugin/src/TranspilerWorker.ts (hunk 3 above)
Comment (dmichon-msft) [reply heft-plugins/heft-typescript-plugin/src/TranspilerWorker.ts:15]: Yep.

File: heft-plugins/heft-typescript-plugin/src/TranspilerWorker.ts (hunk 4 above)
Comment (dmichon-msft) [reply heft-plugins/heft-typescript-plugin/src/TranspilerWorker.ts:19]: Yeah, I've used this a lot when running webpack in worker threads.

**Pull request merged**